How can we install Dragonegg on Mac?

We can install GCC 4.8 via Homebrew

```
brew tap homebrew/versions
brew install gcc48 --with-fortran
```

But this doesn't seem to include GCC plugin headers and so building Dragonegg fails.

Sorry, `gcc48` does actually include the plugin headers - in `lib/gcc/4.8/gcc/x86_64-apple-darwin15.3.0/4.8.5/plugin/include/`.

My problem now is the `libiberty` header. There isn't a package for this as you're supposed to vendor it.

I downloaded the GCC 4.8 source and added the `include` dir (which has `libiberty.h`) as an `-I` on line 85 of the Dragonegg `Makefile`. That gets me a little further, but then I get stuck at:

```
$ GCC=gcc-4.8 CXX=g++-4.8 CC=gcc-4.8 LLVM_CONFIG=../../llvm/bin/llvm-config make
gcc-4.8 --version > /dev/null
../../llvm/bin/llvm-config --version > /dev/null
Compiling Aliasing.cpp
g++-4.8 -c -DTARGET_NAME=\"x86_64-apple-darwin15.3.0\" -I/Users/chrisseaton/Documents/graal/sulong-workspace/sulong/projects/com.oracle.truffle.llvm.tools/tools/dragonegg/dragonegg-3.2.src/include/x86 -I/Users/chrisseaton/Documents/graal/sulong-workspace/sulong/projects/com.oracle.truffle.llvm.tools/tools/dragonegg/dragonegg-3.2.src/include/darwin  -I/Users/chrisseaton/Documents/graal/sulong-workspace/sulong/projects/com.oracle.truffle.llvm.tools/tools/llvm/include  -DNDEBUG -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -fno-rtti -MD -MP -DIN_GCC -DLLVM_VERSION=\"3.2svn\" -DGCC_MAJOR=4 -DGCC_MINOR=8 -DGCC_MICRO=5 -I/Users/chrisseaton/Documents/graal/sulong-workspace/sulong/projects/com.oracle.truffle.llvm.tools/tools/dragonegg/dragonegg-3.2.src/include -isystem/usr/local/Cellar/gcc48/4.8.5/lib/gcc/4.8/gcc/x86_64-apple-darwin15.3.0/4.8.5/plugin/include -I/Users/chrisseaton/Downloads/gcc-4.8.0/include -Wall -Wextra -fvisibility=hidden -I/Users/chrisseaton/Documents/graal/sulong-workspace/sulong/projects/com.oracle.truffle.llvm.tools/tools/llvm/include  -DNDEBUG -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -O3  -fvisibility-inlines-hidden -fno-exceptions -fno-rtti -fno-common -Woverloaded-virtual -Wcast-qual /Users/chrisseaton/Documents/graal/sulong-workspace/sulong/projects/com.oracle.truffle.llvm.tools/tools/dragonegg/dragonegg-3.2.src/src/Aliasing.cpp
In file included from /usr/local/Cellar/gcc48/4.8.5/lib/gcc/4.8/gcc/x86_64-apple-darwin15.3.0/4.8.5/plugin/include/tree.h:27:0,
                 from /Users/chrisseaton/Documents/graal/sulong-workspace/sulong/projects/com.oracle.truffle.llvm.tools/tools/dragonegg/dragonegg-3.2.src/src/Aliasing.cpp:48:
/usr/local/Cellar/gcc48/4.8.5/lib/gcc/4.8/gcc/x86_64-apple-darwin15.3.0/4.8.5/plugin/include/vec.h:230:3: error: template with C linkage
   template <typename, typename, typename> friend struct vec;
   ^
```

So `Aliasing.cpp:48` includes `tree.h`, `tree.h:27` includes `vec.h`, which includes templates, but `Aliasing.cpp` is doing the include in an `extern "C"`, so that doesn't work.

I then tried setting `-DENABLE_BUILD_WITH_CXX`, which seems to turn off the `extern`, but then I get a bunch of errors like this:

```
/Users/chrisseaton/Documents/graal/sulong-workspace/sulong/projects/com.oracle.truffle.llvm.tools/tools/dragonegg/dragonegg-3.2.src/src/Backend.cpp:899:23: error: 'struct cgraph_node' has no member named 'decl'
       target = fnode->decl;
                       ^
```

The whole thing just makes me think Dragonegg isn't as compatible with GCC 4.8 as we thought. @mrigger where did you get the info it worked with GCC 4.8?

Maybe we should go back and keep trying to install GCC 4.6 on the Mac.

--- a/README.md
+++ b/README.md
@@ -25,18 +25,23 @@ Make sure you have GCC-4.6, G++-4.6, and GFortran-4.6 installed. For
 a full list of external dependencies on Ubuntu you can look at our
 [Travis configuration](https://github.com/graalvm/sulong/blob/master/.travis.yml).
 
-On the Mac you can use Homebrew. GCC 4.6 may fail to install, so we'll use
-GCC 4.8 instead.
+On the Mac you can use Homebrew:
 
     brew tap homebrew/versions
-    brew install gcc48 --with-fortran
+    brew install gcc46 --with-fortran
     brew install gmp4
+
+On some versions of Mac OS X, `gcc46` may fail to install with a segmentation
+fault (https://github.com/Homebrew/homebrew-versions/issues/515). A fix for this
+is to `brew edit gcc46` and replace the patch `p0` with
+https://gist.githubusercontent.com/chrisseaton/7008085997269e3478e1/raw/46bff773d37914d6e66d0d8e1cab0d50d77e7280/patch-10.10.diff,
+shasum `51814a0d79a9f21344c76f2d4235b59d9a4bc1601117e8ca5bfabdb82305aad0`.
 
 However you install GCC on the Mac, you may then need to manually link the
 libraries we use into a location where they can be found, as
 `DYLD_LIBRARY_PATH` cannot normally be set on the Mac.
 
-    ln -s /usr/local/Cellar/gcc48/4.8.5/lib/gcc/4.8/libgfortran.3.dylib /usr/local/lib
+    ln -s /usr/local/Cellar/gcc46/4.6.4/lib/gcc/4.6/libgfortran.3.dylib /usr/local/lib
     ln -s /usr/local/Cellar/gmp/6.1.0/lib/libgmp.10.dylib /usr/local/lib
 
 How to get started?

--- a/mx.sulong/dragonegg-mac.patch
+++ b/mx.sulong/dragonegg-mac.patch
@@ -1,0 +1,7 @@
+--- projects/com.oracle.truffle.llvm.tools/tools/dragonegg/dragonegg-3.2.src/Makefile.a	2016-03-27 17:38:21.000000000 +0100
++++ projects/com.oracle.truffle.llvm.tools/tools/dragonegg/dragonegg-3.2.src/Makefile	2016-03-27 17:38:24.000000000 +0100
+@@ -84,3 +84,3 @@
+ 	     -DGCC_MICRO=$(GCC_MICRO) \
+-	     -I$(INCLUDE_DIR) -isystem$(GCC_PLUGIN_DIR)/include
++	     -I$(INCLUDE_DIR) -isystem$(GCC_PLUGIN_DIR)/include -I../../gcc/gcc-4.6.4/include
+ ifdef DISABLE_VERSION_CHECK

--- a/mx.sulong/mx_sulong.py
+++ b/mx.sulong/mx_sulong.py
@@ -169,6 +169,11 @@ def dragonEgg(args=None):
     executeCommand = [getGCC(), "-fplugin=" + _dragonEggPath, '-fplugin-arg-dragonegg-emit-ir']
     return mx.run(executeCommand + args)
 
+def dragonEggGFortran(args=None):
+    """executes GCC Fortran with dragonegg"""
+    executeCommand = [getGFortran(), "-fplugin=" + _dragonEggPath, '-fplugin-arg-dragonegg-emit-ir']
+    return mx.run(executeCommand + args)
+
 def dragonEggGPP(args=None):
     """executes G++ with dragonegg"""
     executeCommand = [getGPP(), "-fplugin=" + _dragonEggPath, '-fplugin-arg-dragonegg-emit-ir']
@@ -200,6 +205,15 @@ def getDefaultGCC():
         return 'gcc46'
     return None
 
+def getDefaultGFortran():
+    # Ubuntu
+    if which('gfortran-4.6') is not None:
+        return 'gfortran-4.6'
+    # Mac
+    if which('gfortran46') is not None:
+        return 'gfortran46'
+    return None
+
 def getDefaultGPP():
     # Ubuntu
     if which('g++-4.6') is not None:
@@ -226,6 +240,17 @@ def getGCC():
         return getDefaultGCC()
     else:
         mx.abort('Could not find a compatible GCC version to execute Dragonegg! Please install gcc-4.6 or another compatible version and specify it in the env file')
+
+def getGFortran():
+    """tries to locate a gfortran version suitable to execute Dragonegg"""
+    specifiedGFortran = getCommand('SULONG_GFORTRAN')
+    if specifiedGFortran is not None:
+        return specifiedGFortran
+    if getDefaultGFortran() is not None:
+        return getDefaultGFortran()
+    else:
+        mx.abort('Could not find a compatible GFortran version to execute Dragonegg! Please install gfortran-4.6 or another compatible version and specify it in the env file')
+
 
 def getGPP():
     """tries to locate a g++ version suitable to execute Dragonegg"""
@@ -613,6 +638,7 @@ mx.update_commands(_suite, {
     'su-clang++' : [compileWithClangPP, ''],
     'su-opt' : [opt, ''],
     'su-gcc' : [dragonEgg, ''],
+    'su-gfortran' : [dragonEggGFortran, ''],
     'su-g++' : [dragonEggGPP, ''],
     'su-travis1' : [travis1, ''],
     'su-travis2' : [travis2, '']

--- a/mx.sulong/mx_sulong.py
+++ b/mx.sulong/mx_sulong.py
@@ -271,6 +271,13 @@ def pullInstallDragonEgg(args=None):
     localPath = pullsuite(toolDir, [url])
     tar(localPath, toolDir)
     os.remove(localPath)
+    if mx.get_os() == 'darwin':
+        gccToolDir = join(_toolDir, "tools/gcc")
+        url = 'http://ftpmirror.gnu.org/gcc/gcc-4.6.4/gcc-4.6.4.tar.gz'
+        localPath = pullsuite(gccToolDir, [url])
+        tar(localPath, gccToolDir)
+        os.remove(localPath)
+        mx.run(['patch', '-p1', _toolDir + 'tools/dragonegg/dragonegg-3.2.src/Makefile', 'mx.sulong/dragonegg-mac.patch'])
     os.environ['GCC'] = getGCC()
     os.environ['CXX'] = getGPP()
     os.environ['CC'] = getGCC()

